mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java

```diff
@@ -550,16 +550,25 @@ private void onHeartbeat(TaskExecutorHeartbeat heartbeat) {
         try {
             final TaskExecutorID taskExecutorID = heartbeat.getTaskExecutorID();
             final TaskExecutorState state = this.executorStateManager.get(taskExecutorID);
-            boolean stateChange = state.onHeartbeat(heartbeat);
-            if (stateChange) {
-                if (state.isAvailable()) {
-                    this.executorStateManager.tryMarkAvailable(taskExecutorID);
+            if (state.getRegistration() == null) {
+                TaskExecutorRegistration registration = this.mantisJobStore.getTaskExecutor(heartbeat.getTaskExecutorID());
```

Comment:
I suppose we should avoid blocking calls at this thread? move to future instead?